articles/content_filtering.md

```diff
@@ -0,0 +1,282 @@
+There are two types of parameter events subscription, one is for system(TimeSource) which is controlled and taken care by rclcpp and rclpy internally. And the other is user application to register user callback to be fired when parameter events come in.
+
+It would be good to create ContentFilteredTopic for each of them. This can make responsibility clear and simple, can avoid the complication about filtering expression merge between user filtering and system filtering.
+
+|  ContentFilterTopic  |  Description  |
+| :--- | :--- |
+|  /parameter_events_system_Node_ID  |  parameter events ContentFilteredTopic for system internal usage, Node ID would be uuid to identify this NodeBase. |
+|  /parameter_events_user_Node_ID  |  parameter events ContentFilteredTopic for user callback  usage, Node ID would be uuid to identify this NodeBase. |
+
+(*) These ContentFilteredTopic should not show up to user. (e.g `ros2 topic list`)
+
+- TimeSource
+  **use_sim_time** event on self-node_base is subscribed internally to check if **use_sim_time** parameter is changed or not. so internally this **use_sim_time** AND self node name always must be in filter_expression and expression_parameters via ContentFilteredTopic if it is supported. see [here](https://github.com/ros2/rclcpp/blob/99286978f92c30fe171313bf0785d6b6272c3257/rclcpp/src/rclcpp/time_source.cpp#L123-L125).
+  **use_sim_time** parameter event must be guaranteed by system to keep that in the filter and parameter expression when using ContentFilteredTopic.
+
+  (Off topic from ContentFilteredTopic)
+  This `use_sim_time` can be supported local callbacks, it does not have to use `/parameter_events` at all.
+
+- User frontend (rclcpp/rclpy)
+  User API will be added to manage filtering configuration, so that user application can set its own filter_expression and expression_parameters. Also using AsyncParametersClient::on_parameter_event, user can take care of the parameter event with user callback only for filtered parameter events. It should be compatible interface for user application even if rmw_implementation does not support ContentFilteredTopic. If rmw_impelementation does not support ContentFilteredTopic, `rcl` will take care of filtering instead based on fitlering expression and parameters.
+
+- Filtering Expression and Parameters
+  User can specify node name and parameter name to filter the parameter evetns.
+  - [T.B.D] This needs to be discussed more, if user should be able to filter arbitrary parameter or not. if to support any expression and parameter, it means that `rcl` needs to have DDS standard expression and parameter filtering function. 
+
+<img src="../img/content_filter/parameter_event_sequence_with_filter.png">
+
+### Interfaces
+
+#### rcl
+
+- add API `is_cft_supported` to allow frontend(rclcpp/rclpy) can know CFT is supported or not.
+  e.g) `rcl_subscription_is_cft_supported` to access `is_cft_supported`.
+- add `rcl_create_cft_subscription` to call `rmw_create_cft_subscription`.

```

Comment:
@fujitatomoya 

Based on your design, I have a question about the function `rmw_create_cft_subscription`.
If using this function, a subscriber can dynamically use `rmw_create_cft_subscription` to set the [topic_reader_](https://github.com/ros2/rmw_connext/blob/25721877d133a6ea09c1f3a864ed892d288fa918/rmw_connext_cpp/src/rmw_subscription.cpp#L224-L225) inside it at any time, which means we need to add an extra lock to protect the `subscriber_info[topic_reader_, read_condition_]` in `rmw_create_cft_subscription` and some other locations like [rmw_connext-wait](https://github.com/ros2/rmw_connext/blob/25721877d133a6ea09c1f3a864ed892d288fa918/rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/wait.hpp#L224-L230).

I was thinking that using `rmw_create_subscription` can create normal topic or content filtered topic based on rmw_subscription_options_t that adding `const char * filter_expression` and `const rcutils_string_array_t * expression_parameters` if CFT is supported. Is it necessary to separate this operation into two actions?

Reply:
I think that is fair and reasonable just to extend the interface to support cft subscription. I will update the design as you suggested.